eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -124,9 +124,17 @@ public void insertAnnotation(IAnnotationModel model, int offset, int length, Use
 
     if (!contribtionAnnotationsEnabled) return;
 
-    if (length < 0) // why is 0 len allowed ?
-    return;
+    for (int i = 0; i < length; i++) {
```

Comment:
My big guess is that if you C&P a long enough text, the receiving side will freeze.

Reply:
What do you think about checking `length > MAX_HISTORY_LENGTH` and just annotating the the last `MAX_HISTORY_LENGTH`s characters? Is a simple optimization that reduces the issue.

Reply:
@tobous How are annotations handled in IntelliJ regarding C&P, e.g one insert = 50 characters ?

And @m273d15 if you want to make mass updates in the annotation model, please calculate the added and removed annotations before you actually replace them. We have a helper function that tries to that at once, i.e only making 1 update call for all affected annotations.

Reply:
In IntelliJ, contribution annotations are handled similarly but not exactly in the way implemented here.

The base approach is the same: All contribution annotations are split into one character wide annotations that do not auto-extend (i.e. don't extend when text is added at either end). See [AnnotationManager.addContributionAnnotation(...)](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationManager.java#L120).
I made this decision to avoid such cases where another user types in the middle of an annotation, thereby extending it if it were a continuously annotated block.

Where the handling seems to differ is that all text added by one activity (like a C&P) is still seen as one annotation, even though it is displayed in many different parts. See [ContributionAnnotation](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/ContributionAnnotation.java). One annotation has a list of `AnnotationRange` objects, which each represent a highlighted section of text.
As a result, all highlightings for the action are bundled and seen as one annotation for the annotation rotation.

In IntelliJ, the performance for this is ok. (I haven't measured the impact.) The funny thing is that, even without Saros, C&P a large chunk (~2000 line) in a VERY large file (~250000 lines) freezes the local editor. Such issues, however, do not occur on the receiving side of such activities. As a result, during a running session, the remove site (which has to deal with all of the annotations but still runs without a hitch) is done far before the sending site.

Reply:
Ok so the logic what is displayed is the same as in Eclipse. I do not see a problem by just using single annotations in Eclipse too. But as I already said, use the bulk method.

Reply:
My last 2 cents. It should not be that hard to modify the history in such way that it stores multiples annotations.

```
List annotationsToRemove = history.getAndRemoveAnnotationsForOldestEntry();
List annotationsToAdd = createAnnotationsForContributionRange(....)
history.addEntry(annotationsToAdd)
model.replaceAnnotations(annotationsToRemove, annotationsToAdd)
```

Reply:
I will modify the history handling and adopt the `AnnotationModelHelper.replaceAnnotationsInModel` method (I think this is the bulk method you talked about).

Reply:
The complexity of the algorithm went up from O(m) to O(m*n). m = number of annotations in the model, n = characters / annotations to add.

Furthermore is it that tragic if multiple annotations are present for a single position ? (In reality it should not happen at all, but it does see the issue in the EditorManager)

As you already know the covered range and if you want to keep the check this algorithm can be optimized to have an average runtime of O(m) again, e.g calculate the range, traverse the annotations only once and collect all that do collide, calculate the colliding annotations afterwards.